Implement equality for toric varieties

paemurru opened the pull request:
The only computationally expensive part is computing the rays of the toric varieties, which are usually already computed anyway.

paemurru committed:
Implement fast equality for toric varieties

The only computationally complex part is computing the rays of the toric varieties, which are usually already computed anyway.

paemurru committed:
Update docs

joschmitt commented:
I vaguely remember that there was a longer discussion about this (parts of it are in #3934, #3946) and eventually it was decided to do things like they are (with the information about `slow_equal` in the docstring). Did something change?

@HereAround @lkastner

HereAround commented:
> I vaguely remember that there was a longer discussion about this (parts of it are in #3934, #3946) and eventually it was decided to do things like they are (with the information about `slow_equal` in the docstring). Did something change?
> 
> @HereAround @lkastner

@paemurru seems to have an increased interest in this feature/functionality. After several discussions more (the last of which I recall did take place before Christmas among @lkastner @paemurru and myself), it was agreed that one could implement such a method - provided interest and time. It seems that @paemurru found both! (Thank you!)

HereAround commented:
@paemurru given the failed required test, I assume this is WIP. Please ping me once ready for review.

lkastner commented:
What is this needed for?

lkastner commented on src/AlgebraicGeometry/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -181,9 +181,16 @@ end
+  f_vector(X) == f_vector(Y) || return false
+
+  # p is a permutation such that the i-th ray of X is the p(i)-th ray of Y
+  p = inv(perm(sortperm(rays(X)))) * perm(sortperm(rays(Y)))
```
Computing the `f_vector` before the rays is absolute overkill. Why not first check equality of the rays? Furthermore computing the `f_vector` also entails computing the maximal cones, so why is this even here?

paemurru replied:
I thought the test `f_vector(X) == f_vector(Y) || return false` was faster than the below. It is not needed otherwise. Do you think it is faster if `f_vector(X) == f_vector(Y) || return false` is removed?

lkastner replied:
The https://en.wikipedia.org/wiki/Upper_bound_theorem makes this computation explode. Because computing the f-vector entails computing all these intersections. I am not familiar with any algorithm that would be able to compute the f_vector without first computing the rays and maximal cones.

lkastner replied:
```
polytope > $c = cyclic(12,15);

polytope > print $c->F_VECTOR;
15 105 455 1365 3003 5005 6420 6330 4690 2478 840 140
polytope > $ff = fan::face_fan($c, $c->REL_INT_POINT);

polytope > print $ff->F_VECTOR;
15 105 455 1365 3003 5005 6420 6330 4690 2478 840 140
polytope > 
```

paemurru replied:
Removed

paemurru commented:
@HereAround Those three test failures are from nightly, those are not required or relevant here! This is not a WIP, that tag can be removed

benlorenz commented on src/AlgebraicGeometry/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -181,9 +181,16 @@ end
+  return Set(map(r -> Set(p.(r)), rows(X))) == Set(rows(Y))
 end
 
 function Base.hash(tv::NormalToricVariety, h::UInt)
```
The hash function needs to be changed to match the equality, so that `x == y` implies `hash(x) == hash(y)`.

paemurru replied:
Changed to match the equality

paemurru replied:
The following would make the hash function have `x == y` if and only if `hash(x) == hash(y)`.

```
function Base.hash(X::NormalToricVariety, h::UInt)
  p = inv(perm(sortperm(rays(X))))
  @inline rows(Z) = [row(maximal_cones(IncidenceMatrix, Z), i) for i in 1:n_maximal_cones(Z)]
  set_of_maximal_cones = Set(map(r -> Set(p.(r)), rows(X)))
  sorted_rays = Vector.([rays(X)[p(i)] for i in 1:n_rays(X)])
  return hash((sorted_rays, set_of_maximal_cones), h)
end
```

But the computation is not instantaneous. I am not sure what the standard is for the hash function speed.

paemurru replied:
Following the recommendation of @simonbrandhorst, I added the above as the hash function, as then the time complexity of finding unique elements of a list of n toric varieties is O(n) instead of O(n^2), for an example improvement.

paemurru commented:
@joschmitt One thing that changed is that I wrote a fast equality check

lkastner commented:
> @joschmitt One thing that changed is that I wrote a fast equality check

It is not fast, since it still potentially involves double description computation. This is why I am asking for a use case. Because in the long run we will have users using this check and then complain about Oscar being slow. Most of the time this equality check can be avoided by studying the underlying mathematics. For example we were able to avoid a lot of equality checks of polyhedra in the git fans that way. Often one is in the situation where one fan is already refining the other fan, e.g. for blow ups. Adding this operator makes this check convenient and seemingly cheap. It is not.

paemurru committed:
Fix equality check, add test, fix hash

Also remove `f_vector`.

paemurru commented:
@lkastner

> It is not fast, since it still potentially involves double description computation.

All the "famous toric varieties" in the documentation already have rays computed, so this does not involve double description computation for them. For any kind of blowups or refining of fans by adding a ray, again the rays are already computed. Often one works with the Cox ring, so the rays need to be computed. I cannot think of a use case where the rays are not already computed.

Would you prefer it if it first checked whether the rays are already computed and if not, it would return an error?

paemurru committed:
Add documentation, improve hash function

paemurru commented:
Should be ready to merge now

lgoettgens commented on src/AlgebraicGeometry/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -181,16 +181,72 @@ end
 end
 
+@doc raw"""
+    hash(X::NormalToricVariety, h::UInt)
```
We usually have no docstring for any `hash` method. As this gives no new insights, I am tempted to not add this here as well.

paemurru replied:
Commented out the docstring. It is still useful information, it gives the insight that this can be used for testing equality, granted that once every 4 billion tries you will have a false positive. This can give great speed improvements for some things as mentioned here https://github.com/oscar-system/Oscar.jl/pull/4509#discussion_r1934039983

lgoettgens commented on src/AlgebraicGeometry/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -181,16 +181,72 @@ end
+Normal toric variety
+
+julia> hash(X) == hash(Y)
+false
```
This doctest can fail randomly due to hash collisions. IMO, we shouldn't add such flaky tests.

paemurru replied:
It would fail once every 4 billion times. I think it does not matter so much. But I commented it out regardless

lgoettgens reviewed (commented):
I cannot say anything about the concrete implementations. But two comments regarding the docstrings below

paemurru committed:
Commented out the hash docstring

benlorenz commented on src/AlgebraicGeometry/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -181,16 +181,50 @@ end
+  p = inv(perm(sortperm(rays(X))))
+  @inline rows(Z) = [row(maximal_cones(IncidenceMatrix, Z), i) for i in 1:n_maximal_cones(Z)]
+  set_of_maximal_cones = Set(map(r -> Set(p.(r)), rows(X)))
+  sorted_rays = Vector.([rays(X)[p(i)] for i in 1:n_rays(X)])
```
```suggestion
  sorted_rays = permuted(collect(rays(X)), inv(p))
```
Maybe use permuted? It would be better to keep the rays as `RayVectors` (since their hash method should ignore scaling).

paemurru replied:
Used `permuted`, as you suggested. I kept as `Vector`, unlike you suggested, since hash of two RayVectors is often the same:

```
julia> ray_generators = [[1, 1], [1, 2]];

julia> max_cones = incidence_matrix([[1, 2]])
1×2 IncidenceMatrix
[1, 2]


julia> X = normal_toric_variety(max_cones, ray_generators)
Normal toric variety

julia> hash(rays(X)[1]) == hash(rays(X)[2])
true
```

benlorenz replied:
I do remember now, for efficiency we consider only the sign vector for the hash and allow fallback to a proper comparison otherwise. (To avoid having to normalize the vectors for the hash. This does not make as much of a difference for rational numbers but it does for more complicated fields)

The ray vectors from a polyhedral fan should be normalized so this should be fine in this case. (In general `RayVector` is not normalized)

benlorenz commented on test/AlgebraicGeometry/ToricVarieties/normal_toric_varieties.jl:
```diff
@@ -41,6 +41,19 @@
+    Z = projective_space(NormalToricVariety, 2)
+    X = domain(blow_up(Z, [1, 1]))
+    Y = domain(blow_up(Z, [1, 2]))
+    @test X != Y
```
To test the hash function maybe create a `Set` of varieties and check the length (or use `unique`)?

paemurru replied:
Great idea. Added

benlorenz commented:
Is it clear (and maybe checked?) that the polyhedral fans for these varieties are pointed (`lineality_dim == 0`)? I could not find anything regarding this in the docs, and otherwise all this breaks down since the rays are not unique.

paemurru commented:
@benlorenz According to the book of Cox–Little–Schenck, polyhedral fans are pointed. At least according to the documentation, toric geometry in OSCAR follows the book of Cox. I am not sure if this is actually checked in the constructor (it should be!)

paemurru committed:
Use permuted

paemurru committed:
Fix inv(p), add hash tests

benlorenz commented on test/AlgebraicGeometry/ToricVarieties/normal_toric_varieties.jl:
```diff
@@ -41,6 +41,26 @@
+    ray_generators = [[1, 1], [1, 2]]
+    max_cones = incidence_matrix([[1, 2]])
+    X = normal_toric_variety(max_cones, ray_generators)
+    @test length(Set(hash.([H, P1 * P1, X]))) == 2
```
```suggestion
    @test length(Set([H, P1 * P1, X])) == 2
```
The julia Set will use hashes (and fall back to the equality), so better use the objects in the Set, otherwise it will fail the test for a hash collision.

paemurru committed:
Update test/AlgebraicGeometry/ToricVarieties/normal_toric_varieties.jl

Co-authored-by: Benjamin Lorenz <benlorenz@users.noreply.github.com>

benlorenz reviewed (approved):
I still don't see why exactly this is necessary but the code seems reasonable now. Probably similar to computing Groebner bases when comparing ideals. So I think we can keep it like this.

But it would still be good to add input checks for toric varieties to make sure the fans are pointed, probably in a separate PR. This is not required for general polyhedral fans in Oscar.

HereAround commented on src/AlgebraicGeometry/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -181,16 +181,52 @@ end
-  error("Equality of normal toric varieties is computationally very demanding. More details in the documentation.")
-end
+@doc raw"""
+    (==)(X::NormalToricVariety, Y::NormalToricVariety)
```
A while back, we took steps to remove "unnecessary" doc-strings. For instance, we have tons of implementation of `dim`, `base_ring`, `gens`, etc. which only differ by signature. Those doc-strings were removed, so that when you search in the online documentation (or issue ? in the REPL), you get fewer results/have an easier time to find what you are really looking for.

I cannot find a description of this decision/design choice in the developer documentation (maybe it is there and I just overlooked it), but the case of equality appears to be similar to me. @thofma ? My suggestion is to remove the `@doc`, and instead describe the equality check directly in the corresponding `.md` file.

thofma replied:
I remember that we removed at lot of redundant docstrings of the form
```
+(x::Bla, y::Bla)

> Returns the sum of x and y
```
(and other functions) for quite a lot signatures. But I can't locate the issue or the PR (only your tangentially related https://github.com/oscar-system/Oscar.jl/issues/1471).

If think for signatures, where there is a chance of ambiguity what it could mean (like in this case), I think it is fine to keep it for now.

Edit: More explanation. In case there is a common supertype (like `RingElem`, `Ring`, `Ideal`), we want to document the mathematical notion directly for this supertype. For example, that `dim(R::Ring)` should return the Krull dimension etc, without putting a `dim` docstring at every possible type, e.g., `dim(::MPolyRing)`, `dim(...)`. But as far as I understand, it is not clear if this makes sense for `==`.

HereAround replied:
Yeah you are right. Thank you for correcting me.

(Added a triage label to my original issue. Maybe this one could be closed, once the decision is described in the style guide...)

HereAround commented on src/AlgebraicGeometry/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -181,16 +181,52 @@ end
+Normal toric variety
+
+julia> H == P1 * P1
+true
```
If you follow my above suggestion, then this could become a CI-test.

HereAround replied:
You can ignore this comment @paemurru . (See above.)

HereAround commented on src/AlgebraicGeometry/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -181,16 +181,52 @@ end
+# outside of hash collisions, `X_1 == X_2` if and only if
+# `hash(X_1) == hash(X_2)`.
+# """
+function Base.hash(X::NormalToricVariety, h::UInt)
```
Are those comments really needed?

We likely (?) do not want to document any hash-function in OSCAR with a designated `@doc`. And as a comment, lines 216 to 222 have - to me - the same content as line 223 alone. Remove the comments?

paemurru replied:
Removed the documentation from the hash function

HereAround commented:
General comment (along our dicussions on slack):

The expectation is that the new equality check could be potentially slow in case polymake has not computed the rays of the toric variety. Let us be careful and reflect this in your changes. I see two ways:

Option 1:
When you describe the equality of toric varieties, please add a warning that informs the users, that this may be potentially slow in the above scenario.

Option 2:
You check in the equality function if the rays are known to `polymake`. For a given toric variety `v`, this should work something like the following:
```julia
"RAYS" in Polymake.list_properties(Oscar.pm_object(polyhedral_fan(v))
```
In case this returns `false`, we default back to the currently existing functionality. (This should also be spelled out in the documentation.)

Note that option 2 requires that you align the hash-function accordingly. In this regard, option 1 is likely easier. So I am leaning towards option 1. But if you prefer option 2, that is also fine with me.

Another useful thing - since we are concerned about the performance of the equality (and hash) operation - could be to subject one CI-test to a time requirement. Here is an in-use example testset:

```julia
  @testset "Speed test for Stanley-Reisner ideal (at most a few seconds)" begin
    success = false
    for i in 1:5
      ntv5 = normal_toric_variety(polarize(polyhedron(Polymake.polytope.rand_sphere(5, 60; seed=42))))
      stats = @timed stanley_reisner_ideal(ntv5)
      duration = stats.time - stats.gctime
      if duration < 10
        success = true
        break
      else
        @warn "Stanley-Reisner ideal took $duration > 10 seconds (i=$i)"
      end
    end
    @test success == true
    ntv5 = normal_toric_variety(polarize(polyhedron(Polymake.polytope.rand_sphere(5, 60; seed=42))))
    @test ngens(stanley_reisner_ideal(ntv5)) == 1648
  end
```

Presumably, any of the tests for == that you wrote should complete in less than 2 (?) seconds?

HereAround reviewed (changes requested):
I left a few comments above @paemurru . Please address those, and then this PR should be good to be merged. Thank you!

paemurru committed:
Remove docstring from the hash function

Also add polymake code to check whether rays are computed into the documentation.

paemurru commented:
I chose Option 1. The slowness in the case the rays have not been computed is already documented and appears in the documentation. Just in case, also mentioned the polymake code to check for this.

About the CI test with a time requirement: this seems a bit overkill to me. It could  happen that it spends a lot of time garbage collecting some previous tests or compiling some functions not before used. And besides, this function here is not that important for time-considerations. There are far more important functions in OSCAR in terms of speed-criticality that have no timed tests whatsoever (saturation, smoothness tests, your own FTheoryTools methods). So I am hesitant to add these here.

paemurru committed:
Speed test

paemurru commented:
Added speed test as @HereAround suggested

HereAround reviewed (approved):
Thank you @paemurru . Looks good to me!
